src/metabase/models/query/permissions.clj

```diff
@@ -59,21 +60,41 @@
 ;;           {:perms/view-data {table-id :unrestricted}}  source-card-read-perms
 ;;
 
-(mu/defn query->source-table-ids :- [:set [:or [:= ::native] ::lib.schema.id/table]]
-  "Return a sequence of all Table IDs referenced by `query`, and/or the ::native keyword for native queries."
+(defn- merge-source-ids
+  "Merge function which takes the union of two sets of IDs, if they are both sets"
+  [val1 val2]
+  (if (and (set? val1) (set? val2))
+    (set/union val1 val2)
+    val2))
```

Comment:
I assume this happens when merging `:native?` keys, where we expect `val1` and `val2` to always both be `true`?

Worth adding a sanity check here to ensure it's safe to ignore `val1`?

Reply:
Yeah, that's the idea. `:native?` should only ever be true. I could change this to `or` two boolean values to make it more explicit.